### PR TITLE
Implement charm

### PR DESCRIPTION
**Pull request by saltiyazan**
# Description
Implementing route53-lego-operator

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        run: tox -e integration
+      - name: Archive Tested Charm
+        uses: actions/upload-artifact@v3
+        if: ${{ github.event_name == 'push' }}
+        with:
+          name: tested-charm
+          path: .tox/**/route53-lego-operator_ubuntu-20.04-amd64.charm
+          retention-days: 5
+  
+  publish-charm:
```
Comment (gruyaume) [inline .github/workflows/main.yaml:58]: Charm publishing shouldn't happen at every push.

Comment (gruyaume) [reply .github/workflows/main.yaml:58]: And make sure the charm is registered and all before merging to main to make sure it gets automatically published.

File: charmcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+parts:
+  charm:
+    build-packages:
+      - libffi-dev
+      - libssl-dev
+      - rustc
+      - cargo
+    charm-binary-python-packages:
+      - psycopg2-binary
```
Comment (gruyaume) [inline charmcraft.yaml:18]: Why is this here?

File: README.md [hunk 3]
```diff
@@ -1,3 +1,62 @@
...
+## OCI Images
+
+This charm uses a [Lego](https://github.com/canonical/lego-rock) image that is built using Rockcraft.
+`ghcr.io/canonical/lego:4.9.1`
+
+## Contributing
+
+Please see the [Juju SDK docs](https://juju.is/docs/sdk) for guidelines on enhancements to this
+charm following best practice guidelines, and
+[CONTRIBUTING.md](https://github.com/canonical/route53-lego-operator/blob/main/CONTRIBUTING.md) for developer
```
Comment (gruyaume) [inline README.md:61]: Where is this CONTRIBUTING.md file?

File: charmcraft.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+type: "charm"
+bases:
+  - build-on:
+      - name: "ubuntu"
+        channel: "20.04"
```
Comment (gruyaume) [inline charmcraft.yaml:5]: Can we use 22.04 here?

File: charmcraft.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    run-on:
+      - name: "ubuntu"
+        channel: "20.04"
```
Comment (gruyaume) [inline charmcraft.yaml:8]: Can we use 22.04 here?

File: metadata.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+name: route53-lego-operator
+
+display-name: Route53 Lego Operator
+
+description: |
+  LEGO operator implementing the provider side of the `tls-certificates`
```
Comment (gruyaume) [inline metadata.yaml:9]: In general try to say ACME instead of LEGO. LEGO is just an implementation of the ACME protocol and could be changed in the future. The same goes with the charm name. It should contain ACME, not LEGO.

File: src/charm.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        """Returns aws hosted zone id from config."""
+        return self.model.config.get("aws_hosted_zone_id")
+
+    @property
+    def _aws_secret_access_key(self):
+        """Returns aws secret access key from config."""
+        return self.model.config.get("aws_secret_access_key")
+
+    @property
+    def _aws_region(self):
```
Comment (gruyaume) [inline src/charm.py:42]: Add return typing. Applies to most methods here.

File: src/charm.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+"""Retrieves certificates from an ACME server using the aws route53 dns provider."""
+
+import logging
+from typing import Dict
+
+from charms.acme_client_operator.v0.acme_client import AcmeClient  # type: ignore[import]
+from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus
+
+# Log messages can be retrieved using juju debug-log
```
Comment (gruyaume) [inline src/charm.py:14]: Not necessary

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: charmcraft.yaml (hunk 4 above)
Comment (ghislainbourgeois) [reply charmcraft.yaml:5]: If this is the case, it would be great to also put 22.04 in all the CI files.

File: config.yaml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    type: string
+    default: ""
+    description: AWS Secret Access Key
+  aws_region:
+    type: string
+    description: AWS Region
+  aws_shared_credentials_file:
+    type: string
+    default: ""
+    description: Path to the file that contains the AWS credentials
```
Comment (ghislainbourgeois) [inline config.yaml:23]: This does not seem used by the code and I assume the file would need to be uploaded to the container. I think we should get rid of it.

File: src/charm.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+    @property
+    def _aws_region(self):
+        """Returns aws region from config."""
+        return self.model.config.get("aws_region")
+
+    @property
+    def _aws_shared_credentials_file(self):
+        """Returns aws shared credentials file path from config."""
+        return self.model.config.get("aws_shared_credentials_file")
```
Comment (ghislainbourgeois) [inline src/charm.py:49]: As above, I think we should get rid of this.

File: src/charm.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+            additional_config["AWS_POLLING_INTERVAL"] = self._aws_polling_interval
+        if self._aws_propagation_timeout:
+            additional_config["AWS_PROPAGATION_TIMEOUT"] = self._aws_propagation_timeout
+        if self._aws_ttl:
+            additional_config["AWS_TTL"] = self._aws_ttl
+        return additional_config
+
+    def _on_config_changed(self, _):
+        """Handles config-changed events."""
+        if not self._aws_shared_credentials_file:
```
Comment (ghislainbourgeois) [inline src/charm.py:94]: As above, we should get rid of this.

File: tests/unit/test_charm.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    ):
+        self.harness.update_config(
+            {
+                "email": "example@email.com",
+                "aws_shared_credentials_file": "./aws-credentials",
+                "aws_region": "dummy region",
+                "aws_hosted_zone_id": "dummy zone id",
+            }
+        )
+        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
```
Comment (ghislainbourgeois) [inline tests/unit/test_charm.py:75]: We are not loading the credentials from the file anywhere, we should get rid of this.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yaml (hunk 1 above)
Comment (saltiyazan) [reply .github/workflows/main.yaml:58]: I followed the same pattern we have in other charms, for example [tls-operator](https://github.com/canonical/tls-certificates-operator/blob/main/.github/workflows/main.yml).
Or am I missing something? :)

File: tests/unit/test_charm.py (hunk 12 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py]: ✅

File: src/charm.py (hunk 11 above)
Comment (saltiyazan) [reply src/charm.py]: ✅

File: src/charm.py (hunk 10 above)
Comment (saltiyazan) [reply src/charm.py]: ✅

File: config.yaml (hunk 9 above)
Comment (saltiyazan) [reply config.yaml]: ✅

File: src/charm.py (hunk 8 above)
Comment (saltiyazan) [reply src/charm.py]: ✅

File: src/charm.py (hunk 7 above)
Comment (saltiyazan) [reply src/charm.py:41]: ✅

File: metadata.yaml (hunk 6 above)
Comment (saltiyazan) [reply metadata.yaml:9]: you are right, the `Lego` specific logic had been moved to the `acme_client` library.
I changed it :) and I will open a PR in `namecheap-lego-operator` targeting the same issue.

File: charmcraft.yaml (hunk 5 above)
Comment (saltiyazan) [reply charmcraft.yaml]: ✅

File: charmcraft.yaml (hunk 4 above)
Comment (saltiyazan) [reply charmcraft.yaml]: ✅

File: README.md (hunk 3 above)
Comment (saltiyazan) [reply README.md:59]: hmm, why wasn't it pushed :O
✅

File: README.md [hunk 13]
```diff
@@ -1,3 +1,62 @@
...
+aws_ttl: <The TTL of the TXT record used for the DNS challenge>
+```
+
+## Relations
+
+`certificates`: `tls-certificates-interface` provider
+
+## OCI Images
+
+This charm uses a [Lego](https://github.com/canonical/lego-rock) image that is built using Rockcraft.
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md]: ✅

File: README.md [hunk 14]
```diff
@@ -1,3 +1,62 @@
...
+  email: <Account email address>
+  aws_access_key_id: <AWS Access Key ID>
+   aws_secret_access_key: <AWS Secret Access Key>
+   aws_region: <AWS Region>
+```
+`juju deploy route53-lego-operator --config <yaml config file>`
+
+Relate it to a `tls-certificates-requirer` charm:
+
+`juju relate route53-lego-operator:certificates tls-certificates-requirer`
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md]: ✅

File: README.md [hunk 15]
```diff
@@ -1,3 +1,62 @@
...
+Deploy route53-lego-operator`:
+
+If you wish to change the default configuration, create a YAML configuration file with fields you would like to change:
+
+
+```yaml
+route53-lego:
+  email: <Account email address>
+  aws_access_key_id: <AWS Access Key ID>
+   aws_secret_access_key: <AWS Secret Access Key>
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md]: ✅

File: README.md [hunk 16]
```diff
@@ -1,3 +1,62 @@
...
+AWS Route53 LEGO operator implements the provider side of the `tls-certificates-interface`
+to provide signed certificates from an ACME servers, using LEGO
+(https://go-acme.github.io/lego).
+It uses the  `acme_client` library to get the certificate from the ACME server.
+
+## Usage
+
+Deploy route53-lego-operator`:
+
+If you wish to change the default configuration, create a YAML configuration file with fields you would like to change:
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md]: ✅

File: README.md [hunk 17]
```diff
@@ -1,3 +1,62 @@
...
+## Description
+
+AWS Route53 LEGO operator implements the provider side of the `tls-certificates-interface`
+to provide signed certificates from an ACME servers, using LEGO
+(https://go-acme.github.io/lego).
+It uses the  `acme_client` library to get the certificate from the ACME server.
+
+## Usage
+
+Deploy route53-lego-operator`:
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md]: ✅

File: README.md [hunk 18]
```diff
@@ -1,3 +1,62 @@
 # route53-lego-operator
 
-This charm implements the acme-client library to retrieve certificates using the ACME protocol for route53 domains.
+## Description
+
+AWS Route53 LEGO operator implements the provider side of the `tls-certificates-interface`
+to provide signed certificates from an ACME servers, using LEGO
+(https://go-acme.github.io/lego).
+It uses the  `acme_client` library to get the certificate from the ACME server.
```
[the remark this answers is not in the document]
Comment (saltiyazan) [reply README.md]: ✅

File: charmcraft.yaml (hunk 2 above)
Comment (saltiyazan) [reply charmcraft.yaml]: ✅

File: lib/charms/acme_client_operator/v0/acme_client.py [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+        """Plugin specific additional configuration for the command.
+
+        Implement this method in your charm to return a dictionary with the plugin specific
+        configuration.
+
+        Returns:
+            dict[str, str]: Plugin specific configuration.
+        """
+
+    def _email_is_valid(self, email: str):
```
Comment (gruyaume) [inline lib/charms/acme_client_operator/v0/acme_client.py:208]: I know that's in the lib but we should add return typing to some of these methods

File: src/charm.py (hunk 7 above)
Comment (gruyaume) [reply src/charm.py:41]: Return typing is still missing in this method

File: tests/unit/test_charm.py [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    def setUp(self):
+        self.harness = Harness(Route53AcmeOperatorCharm)
+        self.harness.set_leader(True)
+        self.harness.set_can_connect("lego", True)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+        self.r_id = self.harness.add_relation("certificates", "remote")
+        self.harness.add_relation_unit(self.r_id, "remote/0")
+
+    def test_given_config_changed_when_email_is_valid_then_status_is_active(self):
```
Comment (gruyaume) [inline tests/unit/test_charm.py:27]: The given and when are inverted. The `when` is the actual action, in this case, `config_changed`. So it should be `given_email_is_valid_when_config_changed_then_status_is_active`. The same issue is present in other tests.

File: tests/unit/test_charm.py [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+
+class TestCharm(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(Route53AcmeOperatorCharm)
+        self.harness.set_leader(True)
+        self.harness.set_can_connect("lego", True)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+        self.r_id = self.harness.add_relation("certificates", "remote")
```
Comment (gruyaume) [inline tests/unit/test_charm.py:24]: I don't think we want to add relations in the setup. It should probably be explicit in the tests that want this relation. And we also probably want to validate behavior for cases where there are no relations.

Comment (gruyaume) [reply tests/unit/test_charm.py:24]: We don't want to abstract everything in the setup. I'd also remove the `set_leader` and `set_connect`. Those are assumptions that won't always be true when we deploy the charm.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/acme_client_operator/v0/acme_client.py (hunk 19 above)
Comment (saltiyazan) [reply lib/charms/acme_client_operator/v0/acme_client.py]: ✅ 
Fixed on the lib side too, please take a look at the pr 
https://github.com/canonical/acme-client-base/pull/6

File: tests/unit/test_charm.py (hunk 20 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py]: ✅

File: tests/unit/test_charm.py (hunk 21 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py]: You are right they were unnecessary.
For validating the behaviour in case of no relation, I'd say that should be done on the lib's side, as this charm is unaware of relations, and it goes to active if its required configuration is provided and passes validation :)
I think the lib doesn't consider that now.

Comment (gruyaume) [reply tests/unit/test_charm.py]: If the behavior is already validated in the lib side that's fine. In all cases I don't think adding relations and setting `can_connect` to `True` should be in the setUp.

**Review by gruyaume — APPROVED** [review]

File: tests/unit/test_charm.py (hunk 21 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py]: Yes, I removed it :)
No, I think the lib actually doesn't validate it, I'm not sure it really needs to, but it's definitely worth investigating.

**Review by ghislainbourgeois — APPROVED** [review]

File: CONTRIBUTING.md [hunk 22]
```diff
@@ -15,19 +15,29 @@ This project uses `tox` for managing test environments. There are some pre-confi
 that can be used for linting and formatting code when you're preparing contributions to the charm:
 
 ```shell
-tox -e fmt           # update your code according to linting rules
 tox -e lint          # code style
 tox -e unit          # unit tests
 tox -e integration   # integration tests
 tox                  # runs 'lint' and 'unit' environments
 ```
```
Comment (sanchezfdezjavier) [inline CONTRIBUTING.md:22]: Have we missed static? Or am I missing something?

File: config.yaml [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    description: Certificate authority server
+    default: "https://acme-v02.api.letsencrypt.org/directory"
+  aws_access_key_id:
+    type: string
+    default: ""
+    description: AWS Access Key ID
+  aws_secret_access_key:
+    type: string
+    default: ""
+    description: AWS Secret Access Key
```
Comment (sanchezfdezjavier) [inline config.yaml:16]: Should we have `defaults` for `aws_access_key_id` and `aws_secret_access_key`? Does this charm even make sense with no authentication?

File: CONTRIBUTING.md (hunk 22 above)
Comment (saltiyazan) [reply CONTRIBUTING.md]: ✅

File: config.yaml (hunk 23 above)
Comment (saltiyazan) [reply config.yaml]: leftovers :)
✅

File: .github/workflows/main.yaml [hunk 24]
```diff
@@ -0,0 +1,83 @@
+name: CI
+
+on:
+  push:
+    branches: ["main"]
```
Comment (Gmerold) [inline .github/workflows/main.yaml:5]: Don't we want to run the CI on every push to any branch? If someone, for some reason, doesn't run the tests locally, he may be really surprised upon PR creation ;)

File: .github/workflows/codeql-analysis.yml [hunk 25]
```diff
@@ -0,0 +1,74 @@
...
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "main" ]
+  schedule:
+    - cron: '25 13 * * 0'
+
+jobs:
+  analyze:
+    name: Analyze
+    runs-on: ubuntu-latest
```
Comment (Gmerold) [inline .github/workflows/codeql-analysis.yml:26]: We use Ubuntu 20.04 in all other places, including `charmcraft.yaml`. Don't we want to use the same version here?

File: .github/workflows/promote.yaml [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+        description: Channel to promote from
+        options:
+          - edge -> beta
+          - beta -> candidate
+          - candidate -> stable
+
+jobs:
+  promote:
+    name: Promote Charm
+    runs-on: ubuntu-latest
```
Comment (Gmerold) [inline .github/workflows/promote.yaml:17]: We use Ubuntu 20.04 in all other places, including `charmcraft.yaml`. Don't we want to use the same version here?

File: README.md [hunk 27]
```diff
@@ -1,3 +1,60 @@
...
+  email: <Account email address>
+  aws_access_key_id: <AWS Access Key ID>
+  aws_secret_access_key: <AWS Secret Access Key>
+  aws_region: <AWS Region>
+```
+`juju deploy route53-acme-operator --config <yaml config file>`
+
+Relate it to a `tls-certificates-requirer` charm:
+
+`juju relate route53-acme-operator:certificates  <tls-certificates-requirer>`
```
Comment (Gmerold) [inline README.md:27]:
```suggestion
`juju relate route53-acme-operator:certificates <tls-certificates-requirer>`
```

File: src/charm.py [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+            self.unit.status = BlockedStatus(
+                "aws-access-key-id, aws-secret-access-key must be set."
+            )
+            return
+        if not self._aws_region:
+            self.unit.status = BlockedStatus("aws-region must be set.")
+            return
+        if not self._aws_hosted_zone_id:
+            self.unit.status = BlockedStatus("aws-hosted-zone-id must be set.")
+            return
```
Comment (Gmerold) [inline src/charm.py:96]: Is there any reason for handling the params separately?

File: tests/unit/test_charm.py [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+                "email": "invalid-email",
+                "aws_access_key_id": "dummy key",
+                "aws_secret_access_key": "dummy access key",
+                "aws_region": "dummy region",
+                "aws_hosted_zone_id": "dummy zone id",
+            }
+        )
+        self.assertEqual(self.harness.model.unit.status, BlockedStatus("Invalid email address"))
+
+    def test_given_credentials_missing_when_config_changed_then_status_is_blocked(
```
Comment (Gmerold) [inline tests/unit/test_charm.py:44]: Using `@parameterized.expand` you could cover all the params combinations with this one test.

File: lib/charms/acme_client_operator/v0/acme_client.py [hunk 30]
```diff
@@ -0,0 +1,244 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# acme_client Library.
+
+This library is designed to enable developers to easily create new charms for the ACME protocol.
+This library contains all the logic necessary to get certificates from an ACME server..
```
Comment (Gmerold) [inline lib/charms/acme_client_operator/v0/acme_client.py:7]:
```suggestion
This library contains all the logic necessary to get certificates from an ACME server.
```

File: tests/integration/test_charm.py [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+import yaml
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test):
```
Comment (sanchezfdezjavier) [inline tests/integration/test_charm.py:19]: Should we add typing?

File: lib/charms/acme_client_operator/v0/acme_client.py [hunk 32]
```diff
@@ -0,0 +1,244 @@
...
+from ops.main import main
+class ExampleAcmeCharm(AcmeClient):
+    def __init__(self, *args):
+        super().__init__(*args, plugin="namecheap")
+        self._server = "https://acme-staging-v02.api.letsencrypt.org/directory"
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+
+    def _on_config_changed(self, _):
+        try:
+            self.update_generic_acme_config(
```
Comment (Gmerold) [inline lib/charms/acme_client_operator/v0/acme_client.py:30]: This method doesn't exist in the base AcmeClient.

File: tests/integration/test_charm.py (hunk 31 above)
Comment (saltiyazan) [reply tests/integration/test_charm.py:19]: on a test function?

File: .github/workflows/main.yaml (hunk 24 above)
Comment (saltiyazan) [reply .github/workflows/main.yaml:5]: Every time I push a change to this PR, the CI is running, I think because of `pull_request` trigger.
shouldn't that be enough?
or do you mean running the CI, before creating the PR?

File: src/charm.py (hunk 28 above)
Comment (saltiyazan) [reply src/charm.py:96]: only the message, should I merge them into one?

File: .github/workflows/main.yaml (hunk 24 above)
Comment (Gmerold) [reply .github/workflows/main.yaml:5]: Yes, I mean before.

Comment (saltiyazan) [reply .github/workflows/main.yaml:5]: I was following the same pattern as some other charms we maintain.
But I will change it :)

File: src/charm.py (hunk 28 above)
Comment (Gmerold) [reply src/charm.py:96]: I think I'd do something similar to what we have done in the `charmed-magma-orchestrator` for the relations. [Click me](https://github.com/canonical/charmed-magma-orchestrator/blob/main/orc8r-orchestrator-operator/src/charm.py#L233)

File: .github/workflows/main.yaml (hunk 24 above)
Comment (Gmerold) [reply .github/workflows/main.yaml:5]: Not sure which ones, but all the `charmed-magma-orchestator` charms run CI on every push to the branch. An example:
```
name: orc8r-accessd

on:
  push:
    paths:
      - "orc8r-accessd-operator/**"
```

File: tests/integration/test_charm.py (hunk 31 above)
Comment (sanchezfdezjavier) [reply tests/integration/test_charm.py:19]: Yeah, forget it, my bad.

File: .github/workflows/codeql-analysis.yml (hunk 25 above)
Comment (sanchezfdezjavier) [reply .github/workflows/codeql-analysis.yml:26]: FYI we have more projects in which we run Codeql with `ubuntu-latest`, I don't say that's the right way of doing it, but just in case.

File: lib/charms/acme_client_operator/v0/acme_client.py (hunk 32 above)
Comment (saltiyazan) [reply lib/charms/acme_client_operator/v0/acme_client.py:30]: ✅

File: lib/charms/acme_client_operator/v0/acme_client.py (hunk 30 above)
Comment (saltiyazan) [reply lib/charms/acme_client_operator/v0/acme_client.py]: ✅

File: README.md (hunk 27 above)
Comment (saltiyazan) [reply README.md]: ✅

File: .github/workflows/promote.yaml (hunk 26 above)
Comment (saltiyazan) [reply .github/workflows/promote.yaml]: ✅

File: lib/charms/acme_client_operator/v0/acme_client.py (hunk 32 above)
Comment (Gmerold) [reply lib/charms/acme_client_operator/v0/acme_client.py:30]: Wait... Did you change anything about that? I can't see any changes ;)

File: .github/workflows/main.yaml (hunk 24 above)
Comment (sanchezfdezjavier) [reply .github/workflows/main.yaml:5]: What about the one in `main.yaml` in `charmed-magma-orchestrator`?
```python
name: main-branch

on:
  pull_request:
    branches:
      - main
  push:
    branches:
      - main
```
My point here is that I don't always want to run the CI on push, I push to the branch with no PR as "cloud storage" to save changes when it's WIP. And when I have something like a "v0" I open a draft PR to start testing and trigger the CI. Don't know if that makes sense.

Comment (Gmerold) [reply .github/workflows/main.yaml:5]: I'm not saying we have to do it. I'm just asking if this was intentional or not.

In any case, in my opinion, ideally, development should be done in the branch, including the CI, and once developer thinks it's ready, PR should be created. Draft PRs I'd save for special occasions (like making changes in the PR specific workflows). But that's just my opinion. While in our case it's not a big problem to do development in PRs, in case of big and active projects, that would generate dozens of PRs and a huge mess.

File: tests/unit/test_charm.py (hunk 29 above)
Comment (saltiyazan) [reply tests/unit/test_charm.py:85]: really nice, thanks for sharing :)
✅

File: lib/charms/acme_client_operator/v0/acme_client.py (hunk 32 above)
Comment (saltiyazan) [reply lib/charms/acme_client_operator/v0/acme_client.py]: oh, didn't push everything sorry for that.
Now it's supposed to be visible :)

File: .github/workflows/codeql-analysis.yml (hunk 25 above)
Comment (saltiyazan) [reply .github/workflows/codeql-analysis.yml:26]: I think @sanchezfdezjavier is right, in many other projects of ours we follow the same practice.
I don't see a contradiction between the code being analysed on `ubuntu-latest` and the charm being built on a specific ubuntu version.
Let me know your thoughts are @Gmerold  :)

File: .github/workflows/main.yaml (hunk 24 above)
Comment (saltiyazan) [reply .github/workflows/main.yaml]: I agree with you @Gmerold, the developer should be able to run CI before creating the PR :)
Done.

File: .github/workflows/codeql-analysis.yml (hunk 25 above)
Comment (Gmerold) [reply .github/workflows/codeql-analysis.yml:26]: No problem. I was just asking.

**Review by Gmerold — APPROVED** [review]

**Review by sanchezfdezjavier — APPROVED** [review]
LGTM
